### PR TITLE
Added Bloom Filter for Orc Files

### PR DESCRIPTION
**Pull request by shalumariamsheji**
### What type of PR is this?

task

### What does this PR do / why do we need it:

Added bloom index to ORC Files for performance improvement while reading.

- Utf8-Bloom Index Write Support is added in ORC files for Real, Date, Timestamp and Char data types for both transactional and non-transactional tables.
- Read flow modified to skip stripes and row groups in ORC based on bloom index.
- Added Timestamp Statistics to Column Statistics.

### Which issue(s) this PR fixes:

None

### Special notes for your reviewers:

### CODE REVIEW

File: presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestOrcWriterOptions.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestOrcWriterOptions.java:1]: huawei copyright header needs to be added

File: presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java [hunk 2]
```diff
@@ -98,19 +98,6 @@ public boolean matches(long numberOfRows, ColumnMetadata<ColumnStatistics> allCo
                 return false;
             }
         }
-
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java:101]: Why OR predicates removed?

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/BloomFilterBuilder.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/BloomFilterBuilder.java:1]: needs the copyright header also

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/NoOpBloomFilterBuilder.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/NoOpBloomFilterBuilder.java:1]: add copyright header

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatistics.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatistics.java:1]: add header copyright

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatisticsBuilder.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatisticsBuilder.java:1]: add copyright header also

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/Utf8BloomFilterBuilder.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/Utf8BloomFilterBuilder.java:1]: add copyright header also

File: presto-orc/src/main/java/io/prestosql/orc/writer/MapColumnWriter.java [hunk 8]
```diff
@@ -199,7 +199,9 @@ public void close()
         ImmutableList.Builder<StreamDataOutput> indexStreams = ImmutableList.builder();
         indexStreams.add(new StreamDataOutput(slice, stream));
         indexStreams.addAll(keyWriter.getIndexStreams(metadataWriter));
+        indexStreams.addAll(keyWriter.getBloomFilters(metadataWriter));
         indexStreams.addAll(valueWriter.getIndexStreams(metadataWriter));
+        indexStreams.addAll(valueWriter.getBloomFilters(metadataWriter));
```
Comment (Nitin-Kashyap) [inline presto-orc/src/main/java/io/prestosql/orc/writer/MapColumnWriter.java:204]: how are blooms used for key and value parts here?

File: presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatistics.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatistics.java:1]: add copyright header

File: presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatisticsBuilder.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
+/*
```
Comment (Nitin-Kashyap) [inline presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatisticsBuilder.java:1]: copyright header need to be added

**Review by Nitin-Kashyap — CHANGES_REQUESTED** [review]
1. Squash multiple commits to one.
2. Add tests in Smoke for the same (i.e. Write table with bloom and read with pushdown filtering)
3. fix the headers in all newly added files in the PR
4. Also update the bloom related writup in hiveConnector md files
5. Update PR details to include changes about Timestamp which are added in the PR

File: presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestOrcWriterOptions.java (hunk 1 above)
Comment (shalumariamsheji) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestOrcWriterOptions.java:1]: fixed

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/BloomFilterBuilder.java (hunk 3 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/BloomFilterBuilder.java:1]: fixed

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/NoOpBloomFilterBuilder.java (hunk 4 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/NoOpBloomFilterBuilder.java:1]: fixed

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatistics.java (hunk 5 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatistics.java:1]: fixed

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatisticsBuilder.java (hunk 6 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/TimestampStatisticsBuilder.java:1]: fixed

File: presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/Utf8BloomFilterBuilder.java (hunk 7 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/metadata/statistics/Utf8BloomFilterBuilder.java:1]: fixed

File: presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatistics.java (hunk 9 above)
Comment (shalumariamsheji) [reply presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatistics.java:1]: fixed

File: presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatisticsBuilder.java (hunk 10 above)
Comment (shalumariamsheji) [reply presto-orc/src/test/java/io/prestosql/orc/metadata/statistics/TestTimestampStatisticsBuilder.java:1]: fixed

File: presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java (hunk 2 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/TupleDomainOrcPredicate.java]: kept back.

File: presto-orc/src/main/java/io/prestosql/orc/writer/MapColumnWriter.java (hunk 8 above)
Comment (shalumariamsheji) [reply presto-orc/src/main/java/io/prestosql/orc/writer/MapColumnWriter.java:204]: Bloom indices are created for both key and value. Support has to be provided for reading them.

**Pull request closed without merging**